### PR TITLE
add masterCA to SA token controller

### PR DESCRIPTION
**Pull request by deads2k**
Fixes https://github.com/openshift/origin/issues/3579

This plumbs a `masterCA` element for `serviceAccountConfig` through the system.  It does not address two additional problems that I discovered:
 1. If we have a external kube and a private key for SA tokens, we don't error or warn in the config validation.  I think we should error, because crazy stuff can happen and there doesn't seem to be a significant upside.
 1. Adding a `masterCA` will not affect SA tokens that already exist, so those will still be unable to verify the api server.

### CODE REVIEW

File: pkg/cmd/server/api/validation/master.go [hunk 1]
```diff
@@ -228,6 +228,12 @@ func ValidateServiceAccountConfig(config api.ServiceAccountConfig, builtInKubern
 		}
 	}
 
+	if len(config.MasterCA) == 0 {
+		validationResults.AddWarnings(fielderrors.NewFieldInvalid("masterCA", "", "master CA information will not be automatically injected into pods, which will prevent verification of the API server from inside a pod"))
```
Comment (liggitt) [inline pkg/cmd/server/api/validation/master.go]: only warn if `builtInKubernetes`?

Comment (liggitt) [inline pkg/cmd/server/api/validation/master.go]: or if `config.PrivateKeyFile`, maybe...

File: pkg/cmd/server/origin/run_components.go [hunk 2]
```diff
@@ -82,10 +83,17 @@ func (c *MasterConfig) RunServiceAccountTokensController() {
 	if err != nil {
 		glog.Fatalf("Error reading signing key for Service Account Token Manager: %v", err)
 	}
+	rootCA := []byte{}
+	if len(c.Options.ServiceAccountConfig.MasterCA) > 0 {
+		rootCA, err = ioutil.ReadFile(c.Options.ServiceAccountConfig.MasterCA)
```
Comment (liggitt) [inline pkg/cmd/server/origin/run_components.go]: run data through `crypto.CertsFromPEM`?

File: pkg/cmd/server/start/master_args.go [hunk 3]
```diff
@@ -254,6 +254,7 @@ func (args MasterArgs) BuildSerializeableMasterConfig() (*configapi.MasterConfig
 		if builtInKubernetes {
 			config.KubeletClientInfo.CA = admin.DefaultRootCAFile(args.ConfigDir.Value())
 			config.KubeletClientInfo.ClientCert = kubeletClientInfo.CertLocation
+			config.ServiceAccountConfig.MasterCA = admin.DefaultRootCAFile(args.ConfigDir.Value())
```
Comment (liggitt) [inline pkg/cmd/server/start/master_args.go]: feels like this should move down to the section where we set `config.ServiceAccountConfig.PublicKeyFiles`... token auth is still applicable even if we're not serving using TLS

Comment (deads2k) [inline pkg/cmd/server/start/master_args.go]: > feels like this should move down to the section where we set config.ServiceAccountConfig.PublicKeyFiles... token auth is still applicable even if we're not serving using TLS

What does this have to do with token auth?  It's about confirming that you're actually talking to the master.  In fact, it only exists if TLS is enabled, right?

File: pkg/cmd/server/origin/run_components.go (hunk 2 above)
Comment (deads2k) [inline pkg/cmd/server/origin/run_components.go]: > run data through crypto.CertsFromPEM to ensure it's actually a cert bundle?

No other CAs do.  Followup?

Comment (liggitt) [inline pkg/cmd/server/origin/run_components.go]: All other CA's do, in the process of building a cert pool

File: pkg/cmd/server/start/master_args.go (hunk 3 above)
Comment (liggitt) [inline pkg/cmd/server/start/master_args.go]: you're right. still feels like it should be in the section with the service account settings, but you're right, it still needs to be guarded by `if args.ListenArg.UseTLS()`

Comment (deads2k) [inline pkg/cmd/server/start/master_args.go]: Since it's tightly coupled to the master using TLS, I'd rather keep it in the top stanza.  It is getting set because we're using TLS and we're starting kube and there's an existing stanza just for that case.

**Comment by deads2k** [comment]
2/3 comments addressed.

File: pkg/cmd/server/api/types.go [hunk 4]
```diff
@@ -346,6 +346,10 @@ type ServiceAccountConfig struct {
 	// Each key is tried in order until the list is exhausted or verification succeeds.
 	// If no keys are specified, no service account authentication will be available.
 	PublicKeyFiles []string
+
+	// MasterCA is the CA for verifying the TLS connection back to the master.  The service account controller will automatically
+	// inject the contents of this file into pods so they can verify connections to the master.
+	MasterCA string
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: I just realized that the OAuth config also needs this (and is currently incorrectly using servingInfo.clientCA instead)... do we want this field replicated in several places, or a single top-level field?

Comment (deads2k) [inline pkg/cmd/server/api/types.go]: > do we want this field replicated in several places, or a single top-level field?

I would replicate it so that the individual configs stay complete in and of themselves, but I don't feel very strongly.

File: pkg/cmd/server/api/v1/types.go [hunk 5]
```diff
@@ -337,6 +337,10 @@ type ServiceAccountConfig struct {
 	// Each key is tried in order until the list is exhausted or verification succeeds.
 	// If no keys are specified, no service account authentication will be available.
 	PublicKeyFiles []string `json:"publicKeyFiles"`
+
+	// MasterCA is the CA for verifying the TLS connection back to the master.  The service account controller will automatically
+	// inject the contents of this file into pods so they can verify connections to the master.
+	MasterCA string `json:"masterCA,omitempty"`
```
Comment (liggitt) [inline pkg/cmd/server/api/v1/types.go]: no omitempty in config, right?

File: pkg/cmd/server/api/types.go (hunk 4 above)
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: ok

File: pkg/cmd/server/start/master_args.go (hunk 3 above)
Comment (liggitt) [inline pkg/cmd/server/start/master_args.go]: ok

**Comment by liggitt** [comment]
update types_test, open doc and ansible issues

File: pkg/cmd/server/origin/run_components.go (hunk 2 above)
Comment (deads2k) [inline pkg/cmd/server/origin/run_components.go]: > All other CA's do, in the process of building a cert pool

done

**Comment by deads2k** [comment]
comments addressed, ansible issue and doc pull opened.

**Comment by liggitt** [comment]
did you add to `GetMasterFileReferences`? I bet that's why the sample didn't get relativized correctly

**Comment by deads2k** [comment]
> did you add to GetMasterFileReferences? I bet that's why the sample didn't get relativized correctly

have now

**Comment by liggitt** [comment]
Fixes #3209

**Comment by liggitt** [comment]
LGTM, [merge]
